Google.Api.Gax.Grpc/ProtobufUtilities.cs

```diff
@@ -0,0 +1,91 @@
+﻿/*
+ * Copyright 2022 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+using Google.Protobuf;
+using Google.Protobuf.Reflection;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Reflection;
+
+namespace Google.Api.Gax.Grpc;
+
+/// <summary>
+/// Utility methods for protobufs. This is deliberately internal; these methods may be effectively exposed
+/// in specific public classes, but only in a way that allows different use cases that happen to have the
+/// same behavior now to be separated later. Code here is also a reasonable candidate to be exposed in Google.Protobuf.
+/// </summary>
+internal static class ProtobufUtilities
+{
+    private const uint UnsignedInt32Zero = 0;
+
+    /// <summary>
+    /// Determines whether the given value is a protobuf default value - i.e. if it's
+    /// null, an empty string, a zero value (integer, numeric or enum) or an empty byte string.
+    /// </summary>
+    internal static bool IsDefaultValue(object value) =>
+        value is null ||
+        // (Comments are the protobuf language keywords)
+        // string
+        value is "" ||
+        // bool
+        value is false ||
+        // double, float
+        value is 0.0f || value is 0.0d ||
+        // int32, sint32, sfixed32, int64, sint64, sfixed64
+        value is 0 || value is 0L ||
+        // uint32, fixed32, uint64, fixed64
+        value is UnsignedInt32Zero || value is 0UL ||
+        // bytes
+        (value is ByteString bs && bs.IsEmpty) ||
+        // Any enum (all protobuf enums have an underlying type of Int32)
+        (value is Enum enumValue && Convert.ToInt32(enumValue) == 0);
+
+    /// <summary>
+    /// Formats a value in a way that is suitable for a header, URL path segment (after URL-encoding), or query parameter. The value
```

Comment:
can't comment here but, where it says `after URL-encoding` did you mean `before URL-encoding`?

These are not encoded, right?

Reply:
Ah, I think you meant it **will** be valid after URL-encoding.

Reply:
Yes, that's right :)